### PR TITLE
docker: prometheus and grafana for development

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
Add prometheus and grafana docker configuration when developing with access to the enterprise repository. This simplifies the setup for developers adding new metrics or even examining events from their development server. The services are disabled by default, but can be enabled by customizing `ENABLED_DOCKER_SERVICES` as described in `config.mk`.

Grafana is provisioned automatically with the Prometheus datasource, as well as the canonical dashboards used with Mattermost. Furthermore, no authentication is required to access Grafana to simplify access from a development environment. The default home dashboard is customized to automatically show links to the provisioned dashboards. Dashboards can be saved, and login remains available via the default `admin/admin` credentials (which initiates a password reset), but no dashboard or system configuration is (currently) persisted if the container is destroyed.

image

Linux requires slightly special handling (well, really it's Docker for MacOS/Windows), in that `host.docker.internal` is the canonical way to resolve the host on MacOS/Windows, but on Linux it's usually sufficient to just use `localhost`. Until https://github.com/docker/for-linux/issues/264 is resolved, this PR includes code to customize the Prometheus configuration to point at the required address for the running platform.

Note that I need some help verifying the Linux changes here, as I don't have access to a development environment running in same.

#### Ticket Link
Fixes: https://mattermost.atlassian.net/browse/MM-31886

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Review by jespino — APPROVED** [review]
LGTM. Thanks I love this.

File: build/docker-compose.common.yml [hunk 1]
```diff
@@ -86,3 +86,17 @@ services:
       - mm-test
     volumes:
      - "./docker/keycloak:/setup"
+  prometheus:
+    image: "prom/prometheus"
+    volumes:
+      - "./docker/prometheus.yml:/etc/prometheus/prometheus${IS_LINUX}.yml"
```
Comment (streamer45) [inline build/docker-compose.common.yml:92]:
```suggestion
      - "./docker/prometheus${IS_LINUX}.yml:/etc/prometheus/prometheus.yml"
```

File: build/docker/prometheus-linux.yml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+global:
+  scrape_interval: 5s
+  evaluation_interval: 60s
+
+scrape_configs:
+  - job_name: 'mattermost'
+    static_configs:
+      - targets: ['localhost:8067']
```
Comment (streamer45) [inline build/docker/prometheus-linux.yml:8]: I am not sure this works. `localhost` from inside the container does not point to the host, hence the need for that special hostname.
I was able to make this work on my Linux setup by using `172.17.0.1` which is the default gateway. More info [here](https://dev.to/natterstefan/docker-tip-how-to-get-host-s-ip-address-inside-a-docker-container-5anh#:~:text=internal%20.&text=On%20Docker%20for%20Linux%2C%20the,you%20are%20using%20default%20networking.).

**Review by streamer45 — COMMENTED** [review]
Amazing work :tada: I've left a couple of suggestions to make this work on Linux.

File: config.mk [hunk 3]
```diff
@@ -5,12 +5,16 @@
...
 #
 # Possible options: mysql, postgres, minio, inbucket, openldap, dejavu,
-# keycloak and elasticsearch
+# keycloak, elasticsearch, prometheus, and grafana.
 #
 # Must be space separated names.
 #
 # Example: mysql postgres elasticsearch
+ifeq ($(BUILD_ENTERPRISE_READY),true)
+ENABLED_DOCKER_SERVICES ?= mysql postgres inbucket prometheus grafana
```
[the remark this answers is not in the document]
Comment (lieut-data) [reply config.mk:14]: Yep, I can get behind that! I'm planning on a blog post around all of this anyway which should help add visibility.

**Comment by lieut-data** [comment]
Thanks for the careful review, @streamer45 -- comments addressed :)

**Review by streamer45 — APPROVED** [review]
Looks great :+1:

**Comment by lieut-data** [comment]
Prapti, would you mind giving this branch a whirl? In principle, you should be able to use this branch and immediately have access to Prometheus on http://localhost:9090 and Grafana on http://localhost:3000. If you install an enterprise license (and restart your server), you should also be able to use the built-in graphs to see metrics for your server.

**Review by prapti — APPROVED** [review]
Thank you for your help on setting up the docker server env variable @lieut-data! Prometheus and grafana dockers get setup properly and are accessible on their respective ports without requiring additional login. Could also use the graphs with the E20 license.
Looks great!

**Pull request merged**